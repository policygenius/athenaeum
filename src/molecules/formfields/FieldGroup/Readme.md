### 3 Field Example:

```jsx
    const selection = [
      { label: 'Option 1',
        value: 1 },
      { label: 'Option 2',
        value: 2 },
      { label: 'Option 3',
        value: 3 }
    ];

    <FieldGroup
      labels={[
        {
          header: 'Test label 1'
        },
        {
          header: 'Test label 2'
        },
        {
          header: 'Test label 3'
        }
      ]}
      input={{}}
      meta={{}}
    >
      <SelectField
        placeholder='Placeholder'
        selectOptions={selection}
      />
      <SelectField
        placeholder='Placeholder'
        selectOptions={selection}
      />
      <SelectField
        placeholder='Placeholder'
        selectOptions={selection}
      />
    </FieldGroup>
```

### 2 Field Example

```jsx
    <FieldGroup
      labels={[
        {
          header: 'Test label 1',
          subLabel: 'Sub label 1'
        },
        {
          header: 'Test label 2',
          subLabel: 'Sub label 2'
        },
      ]}
      input={{}}
      meta={{}}
    >
      <TextField
        placeholder='mm'
        input={{}}
      />
      <TextField
        placeholder='yyyy'
        input={{}}
      />
    </FieldGroup>
```

### Focused Example

```jsx
    <FieldGroup
      labels={[
        {
          header: 'Test label 1',
        },
      ]}
      input={{}}
      meta={{
        active: true,
      }}
    >
      <TextField
        placeholder='mm'
        input={{}}
      />
      <TextField
        placeholder='yyyy'
        input={{}}
      />
    </FieldGroup>
```

### Error Example

```jsx
    <FieldGroup
      labels={[
        {
          header: 'Test label 1',
        },
        {
          header: 'Test label 2',
        },
        {
          header: 'Test label 3',
        },
      ]}
      input={{}}
      meta={{
        visited: true,
        error: 'Some error',
      }}
    >
      <TextField
        placeholder='mm'
        input={{}}
      />
      <TextField
        placeholder='dd'
        input={{}}
      />
      <TextField
        placeholder='yyyy'
        input={{}}
      />
    </FieldGroup>
```
