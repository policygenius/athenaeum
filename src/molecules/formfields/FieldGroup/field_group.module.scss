@import '../shared/formfields.module.scss';

.field-group {
  border: 1px solid color('neutral-4');
}

.label-wrapper {
  padding: ru(.5) 0;
  border-bottom: 1px solid color('neutral-4');
}

.label {
  @extend %label;
  padding: 0 ru(.75);
}

.field {
  padding: 0;
  border-right: 1px solid color('neutral-4');

  &:last-child {
    border-right: 0;
  }
}

/* --  state:focus  -- */
.focused {
  border: 1px solid color('secondary-2');
}

/* --  state:error  -- */
.hasError {
  border: 1px solid color('primary-1');
}
