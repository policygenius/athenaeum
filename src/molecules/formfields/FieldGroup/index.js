/* eslint-disable jsx-a11y/label-has-for */
import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import Col from 'atoms/Layout/Col';
import Layout from 'atoms/Layout';
import Text from 'atoms/Text';
import ErrorMessage from 'atoms/ErrorMessage';

import styles from './field_group.module.scss';

const renderLabels = (labels) => {
  const labelCols = 12 / labels.length;

  return (
    <Layout
      smallCols={[ labelCols ]}
      fullwidth
      className={styles['label-wrapper']}
    >
      {
        labels.map((label, idx) =>
          <div
            key={`field--label-${idx}`}
            className={styles['label']}
          >
            <label>{label.header}</label>
            {
              label.subLabel &&
                <Text
                  size={10}
                  font='b'
                >
                  {label.subLabel}
                </Text>
            }
          </div>
        )
      }
    </Layout>
  );
};

const wrapChild = child =>
  <Col className={styles['field']}>
    {
      React.cloneElement(
        child,
        {
          noBaseStyle: true,
        }
      )
    }
  </Col>;

function FieldGroup(props) {
  const {
    children,
    className,
    labels,
    input,
    meta,
  } = props;

  const childCols = 12 / children.length;

  const showErrorMessage = (meta.visited && !meta.active) || meta.submitFailed;
  const message = meta && showErrorMessage && (meta.error || meta.warning);

  const classes = [
    styles['field-group'],
    meta && meta.active && styles['focused'],
    meta && showErrorMessage && meta.error && !meta.active && styles['hasError'],
    className,
  ];

  return (
    <div>
      <div
        className={classnames(...classes)}
        onBlur={(e) => {
          if (!e.relatedTarget) input.onBlur();
        }}
        onFocus={input && input.onFocus}
      >
        { renderLabels(labels) }

        <Layout
          smallCols={[ childCols ]}
          fullwidth
        >
          {
            React.Children.map(children, wrapChild)
          }
        </Layout>
      </div>

      <ErrorMessage
        condition={!!message}
        message={message}
      />
    </div>
  );
}


FieldGroup.propTypes = {
  /**
   * This prop will add a new className to any inherent classNames
   * provided in the component's index.js file.
   */
  className: PropTypes.string,
  /**
   * Labels to place above each field. The order of the labels must match the order of the child fields.
   */
  labels: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string,
    subLabel: PropTypes.string,
  })),
  /**
   * Meta object is passed from reduxForm
   */
  meta: PropTypes.object,
  /**
   * Input object is passed from reduxForm
   */
  input: PropTypes.object
};

FieldGroup.defaultProps = {
  // Place any default props here.
};

export default FieldGroup;
