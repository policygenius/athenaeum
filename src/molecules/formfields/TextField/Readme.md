Text field without label:

```jsx
    class TextFieldExample extends React.Component {
      constructor() {
        super();

        this.state = {
          value: ''
        };

        this.onChange = this.onChange.bind(this);
      }

      onChange(event) {
        this.setState({ value: event.target.value });
      }

      render() {
        return (
          <TextField
            placeholder='Enter your first name here'
            input={{
              value: this.state.value,
              onChange: this.onChange
            }}
          />
        )
      }
    }

    <TextFieldExample />
```
Text field:

```jsx
    class TextFieldExample extends React.Component {
      constructor() {
        super();

        this.state = {
          value: ''
        };

        this.onChange = this.onChange.bind(this);
      }

      onChange(event) {
        this.setState({ value: event.target.value });
      }

      render() {
        return (
          <TextField
            label='First Name'
            placeholder='Enter your first name here'
            tooltip={"I'm a tooltip"}
            input={{
              value: this.state.value,
              onChange: this.onChange
            }}
            subLabel='Optional subheader'
          />
        )
      }
    }

    <TextFieldExample />
```

Text field with focus:

```jsx
  <TextField
    label='First Name'
    placeholder='Enter your first name'
    tooltip={"I'm a tooltip"}
    input={{
      value: state.value,
      onChange: (event) => setState({ value: event.target.value })
    }}
    meta={{
      active: true
    }}
  />
```

Text field with error:

```jsx
  <TextField
    label='First Name'
    placeholder='Enter your first name'
    tooltip={"I'm a tooltip"}
    input={{
      value: state.value,
      onChange: (event) => setState({ value: event.target.value })
    }}
    meta={{
      error: 'Maecenas ullamcorper turpis quis elit volutpat',
      touched: true
    }}
  />
```

Text field - phone number example:

```jsx
  <TextField
    label='Phone Number'
    placeholder='In case of emergency'
    tooltip={() => alert("I'm a tooltip")}
    input={{
      value: state.value,
      onChange: event => setState({ value: event.target.value })
    }}
    mask='999-999-9999'
    maskChar='_'
  />
```
Text field - prefix example:

```jsx
  <TextField
    label='Income'
    placeholder='Eg. 200,000'
    tooltip={() => alert("I'm a tooltip")}
    input={{
      value: state.value,
      onChange: event => setState({ value: event.target.value })
    }}
    prefix='$'
  />
```
Text field - postfix example:

```jsx
  <TextField
    label='Weight'
    placeholder='Enter your weight'
    tooltip={() => alert("I'm a tooltip")}
    input={{
      value: state.value,
      onChange: event => setState({ value: event.target.value })
    }}
    postfix='lbs.'
  />
```

Text field - secure:

```jsx
  <TextField
    label='Phone Number'
    placeholder='In case of emergency'
    input={{
      value: state.value,
      onChange: event => setState({ value: event.target.value })
    }}
    mask='999-999-9999'
    maskChar='_'
    secure
  />
```
