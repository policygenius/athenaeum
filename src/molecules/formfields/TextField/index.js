import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import styles from '../shared/formfields.module.scss';

function TextField( props ) {
  const {
    className,
    placeholder,
    label,
    input,
    htmlFor,
    meta,
  } = props;

  return (
    <div className={classnames(styles.textfield, { [styles.focused]: meta && meta.active }, className)}>
      {
        label &&
        <label
          className={styles.label}
          htmlFor={htmlFor}
        >
          { label }
        </label>
      }

      <input
        className={styles.input}
        type='text'
        placeholder={placeholder}
        {...input}
      />
    </div>
  );
}

TextField.propTypes = {
  /**
   * Label is optional. If not provided, component will reorganize accordingly.
   */
  label: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object
  ]),
  /**
   * `for` prop on label
   */
  htmlFor: PropTypes.string,

  /**
   * Will append new classname to classSet
   */
  className: PropTypes.string,

  /**
   * placeholder text for text field
   */
  placeholder: PropTypes.string,

  /**
   * object with all necessary props for input
   */
  input: PropTypes.object,
  /**
   * The props under the meta key are metadata about the state of this field that `redux-form` tracks.
   */
  meta: PropTypes.object,
};

TextField.defaultProps = {
  placeholder: 'Placeholder'
};

export default TextField;
