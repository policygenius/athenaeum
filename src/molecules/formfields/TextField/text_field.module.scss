@import '../shared/formfields.module.scss';

.text-field {
  @extend %form-field;
}

.header {
  padding: ru(.5) ru(.75);

  border-bottom: $border;
}

.label-wrapper {
  display: flex;
  align-items: center;
  width: 100%;
}

.icon-lock {
  width: ru(.75);
  height: ru(.75);
  margin-left: auto;
}

.label {
  @extend %label;
  padding: 0;
}

.input {
  @extend %field;
  @include placeholder();
  width: 100%;
  color: color('primary-3');
  font-weight: 400;
}

.tooltip {
  margin-left: auto;
}

.tooltip-icon {
  @extend %tooltip-icon;
}

.input-tooltip {
  position: absolute;
  right: ru(0.5);
  bottom: ru(0);
}

.input-wrapper {
  position: relative;
}

@mixin text-field-fix() {
  display: flex;
  align-items: center;
  flex: 1 0 auto;
  padding: ru(.5) ru(1);
  text-align: center;
  background-color: color('neutral-5');
  color: color('primary-3');
}

.prefix {
  display: flex;

  &:before {
    @include text-field-fix();
    content: attr(data-prefix);
    border-right: 1px solid color('neutral-4');
  }
}

.postfix {
  display: flex;

  &:after {
    @include text-field-fix();
    content: attr(data-postfix);
    border-left: 1px solid color('neutral-4');
  }
}

input[type='text'] {
  outline: none;
}

/* --  state:focus  -- */
.focused {
  @extend %focused;
}

/* --  state:error  -- */
.hasError {
  @extend %hasError;
}
