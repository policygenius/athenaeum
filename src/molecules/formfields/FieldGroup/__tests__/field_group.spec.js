/* eslint-disable no-extra-parens */
import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import FieldGroup from 'molecules/formfields/FieldGroup';
import TextField from 'molecules/formfields/TextField';

describe('<FieldGroup />', () => {
  let component;

  beforeEach(() => {
    component =
      (<FieldGroup
        labels={[
          {
            label: 'Test',
          },
          {
            label: 'Another test',
          },
        ]}
        input={{}}
        meta={{}}
      >
        <TextField
          placeholder='mm'
          input={{}}
        />
        <TextField
          placeholder='yyyy'
          input={{}}
        />
      </FieldGroup>)
    ;
  });

  it('renders correctly', () => {
    const wrapper = mount(component);

    expect(wrapper.type()).toEqual(FieldGroup);
  });

  it('renders correctly', () => {
    const actual = renderer.create(component).toJSON();

    expect(actual).toMatchSnapshot();
  });
});
