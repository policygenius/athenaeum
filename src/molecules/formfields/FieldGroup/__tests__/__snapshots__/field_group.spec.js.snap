// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FieldGroup /> renders correctly 1`] = `
<div>
  <div
    className="field-group"
    onBlur={[Function]}
  >
    <div
      className="layout label-wrapper"
    >
      <div
        className="col col-sm-6 fullwidth border-neutral-6"
      >
        <div
          className="label"
        >
          <label />
        </div>
      </div>
      <div
        className="col col-sm-6 fullwidth border-neutral-6"
      >
        <div
          className="label"
        >
          <label />
        </div>
      </div>
    </div>
    <div
      className="layout"
    >
      <div
        className="col col-sm-6 fullwidth border-neutral-6 field"
      >
        <div>
          <div
            className=""
          >
            <span
              className="input-wrapper"
            >
              <input
                className="input"
                placeholder="mm"
                type="text"
              />
            </span>
          </div>
          <span />
        </div>
      </div>
      <div
        className="col col-sm-6 fullwidth border-neutral-6 field"
      >
        <div>
          <div
            className=""
          >
            <span
              className="input-wrapper"
            >
              <input
                className="input"
                placeholder="yyyy"
                type="text"
              />
            </span>
          </div>
          <span />
        </div>
      </div>
    </div>
  </div>
  <span />
</div>
`;
